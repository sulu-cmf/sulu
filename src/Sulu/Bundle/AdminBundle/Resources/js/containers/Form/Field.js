// @flow
import React from 'react';
import {observable} from 'mobx';
import type {SchemaEntry} from '../../stores/ResourceStore/types';
import fieldRegistry from './registries/FieldRegistry';
import fieldStyles from './field.scss';

type Props = {
    name: string,
    value?: mixed,
    schema: SchemaEntry,
    onChange: (string, mixed) => void,
    locale?: observable,
};

export default class Field extends React.PureComponent<Props> {
    handleChange = (value: mixed) => {
        const {name, onChange} = this.props;

        onChange(name, value);
    };

    render() {
        const {
            value,
            locale,
            schema,
        } = this.props;
        const {label, type} = schema;
        const FieldType = fieldRegistry.get(type);

        return (
            <div>
                <label className={fieldStyles.label}>{label}</label>
                <FieldType
                    onChange={this.handleChange}
                    value={value}
                    locale={locale}
                />
            </div>
        );
    }
}
