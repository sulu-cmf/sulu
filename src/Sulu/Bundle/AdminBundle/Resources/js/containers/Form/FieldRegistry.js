// @flow
import type {ComponentType} from 'react';

class FieldRegistry {
    fields: {[string]: ComponentType<*>};

    constructor() {
        this.clear();
    }

    clear() {
        this.fields = {};
    }

    add(name: string, field: ComponentType<*>) {
        if (name in this.fields) {
            throw new Error('The key "' + name + '" has already been used for another field');
        }

        this.fields[name] = field;
    }

    get(name: string) {
        return this.fields[name];
    }
}

export default new FieldRegistry();
