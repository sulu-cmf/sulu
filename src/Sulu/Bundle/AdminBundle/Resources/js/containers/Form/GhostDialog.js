// @flow
import React from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import Dialog from '../../components/Dialog';
import SingleSelect from '../../components/SingleSelect';
import {translate} from '../../utils/Translator';
import ghostDialogStyles from './ghostDialog.scss';

type Props = {
    locales: Array<string>,
    onCancel: () => void,
    onConfirm: (locale: string) => void,
    open: boolean,
};

@observer
export default class GhostDialog extends React.Component<Props> {
    @observable selectedLocale: string;

    constructor(props: Props) {
        super(props);

        this.selectedLocale = this.props.locales[0];
    }

    handleCancel = () => {
        this.props.onCancel();
    };

    handleConfirm = () => {
        this.props.onConfirm(this.selectedLocale);
    };

    @action handleLocaleChange = (locale: string | number) => {
        if (typeof locale !== 'string') {
            throw new Error('Only strings are accepted as locales! This should not happen and is likely a bug.');
        }

        this.selectedLocale = locale;
    };

    render() {
        const {
            locales,
            open,
        } = this.props;

        return (
            <Dialog
                cancelText={translate('sulu_admin.no')}
                onCancel={this.handleCancel}
                confirmText={translate('sulu_admin.yes')}
                onConfirm={this.handleConfirm}
                open={open}
                title={translate('sulu_admin.ghost_dialog_title')}
            >
                <div className={ghostDialogStyles.ghostDialog}>
                    <p>{translate('sulu_admin.ghost_dialog_description')}</p>
                    <label className={ghostDialogStyles.label}>{translate('sulu_admin.choose_language')}</label>
                    <div className={ghostDialogStyles.localeSelect}>
                        <SingleSelect onChange={this.handleLocaleChange} value={this.selectedLocale}>
                            {locales.map((locale) => (
                                <SingleSelect.Option key={locale} value={locale}>
                                    {locale}
                                </SingleSelect.Option>
                            ))}
                        </SingleSelect>
                    </div>
                </div>
            </Dialog>
        );
    }
}
