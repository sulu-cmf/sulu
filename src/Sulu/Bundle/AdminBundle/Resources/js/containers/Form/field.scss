@import '../Application/colors.scss';

$exceptionColor: $white;
$exceptionBackgroundColor: $roman;
$exceptionBorderColor: $persianRed;
$exceptionBorderColor: $persianRed;

.field-exception {
    color: $exceptionColor;
    background: $exceptionBackgroundColor;
    border: 1px solid $exceptionBorderColor;
    padding: 0 20px;
    border-radius: 3px;
    margin-bottom: 20px;
}
