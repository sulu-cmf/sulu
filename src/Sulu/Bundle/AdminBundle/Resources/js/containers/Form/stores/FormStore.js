// @flow
import {action, computed, observable} from 'mobx';
import type {IObservableValue} from 'mobx'; // eslint-disable-line import/named
import ResourceStore from '../../../stores/ResourceStore';
import type {Schema} from '../types';
import metadataStore from './MetadataStore';

function addSchemaProperties(data: Object, key: string, schema: Schema) {
    const type = schema[key].type;

    if (type !== 'section') {
        data[key] = null;
    }

    const items = schema[key].items;

    if (type === 'section' && items) {
        Object.keys(items)
            .reduce((object, childKey) => addSchemaProperties(data, childKey, items), data);
    }

    return data;
}

export default class FormStore {
    resourceStore: ResourceStore;
    schema: Schema;
    @observable schemaLoading: boolean = true;

    constructor(resourceStore: ResourceStore) {
        this.resourceStore = resourceStore;

        metadataStore.getSchema(this.resourceStore.resourceKey)
            .then(action((schema) => {
                this.changeSchema(schema);
                this.schemaLoading = false;
            }));
    }

    @computed get loading(): boolean {
        return this.resourceStore.loading || this.schemaLoading;
    }

    @computed get data(): Object {
        return this.resourceStore.data;
    }

    save() {
        this.resourceStore.save();
    }

    set(name: string, value: mixed) {
        this.resourceStore.set(name, value);
    }

    @computed get locale(): ?IObservableValue<string> {
        return this.resourceStore.locale;
    }

    changeSchema(schema: Schema) {
        this.schema = schema;
        const schemaFields = Object.keys(schema).reduce((data, key) => addSchemaProperties(data, key, schema), {});

        this.resourceStore.data = {...schemaFields, ...this.resourceStore.data};
    }
}
