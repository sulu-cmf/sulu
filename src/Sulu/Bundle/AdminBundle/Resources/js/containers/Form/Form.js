// @flow
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import React from 'react';
import Loader from '../../components/Loader';
import Renderer from './Renderer';
import FormStore from './stores/FormStore';

type Props = {
    store: FormStore,
    onSubmit: (action: ?string) => void,
};

@observer
export default class Form extends React.Component<Props> {
    @observable showAllErrors = false;

    /** @public */
    @action submit = (action: ?string) => {
        this.showAllErrors = true;
        this.props.onSubmit(action);
    };

    handleChange = (name: string, value: mixed) => {
        this.props.store.change(name, value);
    };

    handleFieldFinish = () => {
        this.props.store.validate();
    };

    render() {
        const {store} = this.props;

        return store.loading
            ? <Loader />
            : (
                <form>
                    <Renderer
                        data={store.data}
                        errors={store.errors}
                        locale={store.locale}
                        onChange={this.handleChange}
                        onFieldFinish={this.handleFieldFinish}
                        schema={store.schema}
                        showAllErrors={this.showAllErrors}
                    />
                </form>
            );
    }
}
