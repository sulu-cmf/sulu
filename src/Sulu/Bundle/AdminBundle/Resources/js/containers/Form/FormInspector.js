// @flow
import type {IObservableValue} from 'mobx'; // eslint-disable-line import/named
import pointer from 'jsonpointer';
import FormStore from './stores/FormStore';

export default class FormInspector {
    formStore: FormStore;

    constructor(formStore: FormStore) {
        this.formStore = formStore;
    }

    get resourceKey(): string {
        return this.formStore.resourceKey;
    }

    get locale(): ?IObservableValue<string> {
        return this.formStore.locale;
    }

    get id(): ?string | number {
        return this.formStore.id;
    }

    getValueByPath(path: string): mixed {
        return pointer.get(this.formStore.data, path);
    }
}
