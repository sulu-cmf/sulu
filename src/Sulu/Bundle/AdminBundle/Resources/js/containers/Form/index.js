// @flow
import Form from './Form';
import FormInspector from './FormInspector';
import fieldRegistry from './registries/FieldRegistry';
import FormStore from './stores/FormStore';
import Selection from './fields/Selection';
import SingleSelection from './fields/SingleSelection';
import Checkbox from './fields/Checkbox';
import ColorPicker from './fields/ColorPicker';
import DatePicker from './fields/DatePicker';
import Email from './fields/Email';
import Input from './fields/Input';
import Number from './fields/Number';
import PasswordConfirmation from './fields/PasswordConfirmation';
import Phone from './fields/Phone';
import SingleSelect from './fields/SingleSelect';
import ResourceLocator from './fields/ResourceLocator';
import Renderer from './Renderer';
import TextArea from './fields/TextArea';
import TextEditor from './fields/TextEditor';
import Time from './fields/Time';
import type {Schema, Types} from './types';

export {
    fieldRegistry,
    Selection,
    Checkbox,
    ColorPicker,
    DatePicker,
    Email,
    Input,
    FormInspector,
    FormStore,
    Number,
    PasswordConfirmation,
    Phone,
    Renderer,
    ResourceLocator,
    SingleSelect,
    SingleSelection,
    TextArea,
    TextEditor,
    Time,
};
export type {Schema, Types};
export default Form;
