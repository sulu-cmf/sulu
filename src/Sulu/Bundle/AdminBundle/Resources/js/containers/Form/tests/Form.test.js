// @flow
import React from 'react';
import {observable} from 'mobx';
import {mount, render, shallow} from 'enzyme';
import Form from '../Form';
import ResourceStore from '../../../stores/ResourceStore';
import FormStore from '../stores/FormStore';
import metadataStore from '../stores/MetadataStore';

jest.mock('../../../utils/Translator', () => ({
    translate: (key) => key,
}));

jest.mock('../registries/FieldRegistry', () => ({
    get: jest.fn((type) => {
        switch (type) {
            case 'block':
                return require('../../../containers/FieldBlocks').default;
            case 'text_line':
                return require('../../../components/Input').default;
        }
    }),
    getOptions: jest.fn().mockReturnValue({}),
}));

jest.mock('../stores/FormStore', () => jest.fn(function(resourceStore) {
    this.id = resourceStore.id;
    this.resourceKey = resourceStore.resourceKey;
    this.data = resourceStore.data;
    this.locale = resourceStore.locale;
    this.loading = resourceStore.loading;
    this.validate = jest.fn();
    this.schema = {};
    this.set = jest.fn();
    this.change = jest.fn();
    this.finishField = jest.fn();
    this.isFieldModified = jest.fn();
    this.copyFromLocale = jest.fn();
    this.getValueByPath = jest.fn();
}));

jest.mock('../../../stores/ResourceStore', () => jest.fn(function(resourceKey, id, observableOptions = {}) {
    this.resourceKey = resourceKey;
    this.id = id;
    this.data = {};
    this.locale = observableOptions.locale;
    this.setLocale = jest.fn((locale) => this.locale.set(locale));
    this.loading = false;
}));

jest.mock('../stores/MetadataStore', () => ({
    getSchema: jest.fn(),
}));

test('Should render form using renderer', () => {
    const submitSpy = jest.fn();
    const store = new FormStore(new ResourceStore('snippet', '1'));

    const form = render(<Form onSubmit={submitSpy} store={store} />);
    expect(form).toMatchSnapshot();
});

test('Should call onSubmit callback', () => {
    const submitSpy = jest.fn();
    const store = new FormStore(new ResourceStore('snippet', '1'));
    metadataStore.getSchema.mockReturnValue({});

    const form = mount(<Form onSubmit={submitSpy} store={store} />);

    form.instance().submit();

    expect(submitSpy).toBeCalled();
});

test('Should validate form when a field has finished being edited', () => {
    const store = new FormStore(new ResourceStore('snippet', '1'));
    metadataStore.getSchema.mockReturnValue({});

    const form = mount(<Form onSubmit={jest.fn()} store={store} />);

    form.find('Renderer').prop('onFieldFinish')();

    expect(store.validate).toBeCalledWith();
});

test('Should validate form before calling finish handlers when a field has finished being edited', () => {
    const handler1 = jest.fn(() => {
        expect(validateCalled).toEqual(true);
    });
    const store = new FormStore(new ResourceStore('snippet', '1'));
    metadataStore.getSchema.mockReturnValue({});

    const form = mount(<Form onSubmit={jest.fn()} store={store} />);
    form.instance().formInspector.addFinishFieldHandler(handler1);

    let validateCalled = false;
    store.validate.mockImplementation(() => validateCalled = true);
    form.find('Renderer').prop('onFieldFinish')();
});

test('Call finish handlers with dataPath and schemaPath when a section field has finished being edited', () => {
    const handler1 = jest.fn();
    const handler2 = jest.fn();

    const store = new FormStore(new ResourceStore('snippet', '1'));
    // $FlowFixMe
    store.schema = {
        highlight: {
            items: {
                title: {
                    type: 'text_line',
                },
            },
            type: 'section',
        },
    };
    const form = mount(<Form onSubmit={jest.fn()} store={store} />);
    form.instance().formInspector.addFinishFieldHandler(handler1);
    form.instance().formInspector.addFinishFieldHandler(handler2);

    form.find('Field[name="title"] Input').prop('onFinish')();
    expect(handler1).toHaveBeenLastCalledWith('/title', '/highlight/items/title');
    expect(handler2).toHaveBeenLastCalledWith('/title', '/highlight/items/title');
});

test('Call finish handlers with dataPath and schemaPath when a field has finished being edited', () => {
    const handler1 = jest.fn();
    const handler2 = jest.fn();

    const store = new FormStore(new ResourceStore('snippet', '1'));
    // $FlowFixMe
    store.schema = {
        article: {
            type: 'text_line',
        },
    };
    const form = mount(<Form onSubmit={jest.fn()} store={store} />);
    form.instance().formInspector.addFinishFieldHandler(handler1);
    form.instance().formInspector.addFinishFieldHandler(handler2);

    form.find('Field[name="article"] Input').prop('onFinish')();
    expect(handler1).toHaveBeenLastCalledWith('/article', '/article');
    expect(handler2).toHaveBeenLastCalledWith('/article', '/article');
});

test('Call finish handlers with dataPath and schemaPath when a block field has finished being edited', () => {
    const handler1 = jest.fn();
    const handler2 = jest.fn();

    const resourceStore = new ResourceStore('snippet', '1');
    resourceStore.data = {
        block: [
            {
                text: 'Test',
                type: 'default',
            },
        ],
    };

    const store = new FormStore(resourceStore);
    // $FlowFixMe
    store.schema = {
        block: {
            type: 'block',
            types: {
                default: {
                    form: {
                        text: {
                            type: 'text_line',
                        },
                    },
                    title: 'Default',
                },
            },
        },
    };

    const form = mount(<Form onSubmit={jest.fn()} store={store} />);
    form.instance().formInspector.addFinishFieldHandler(handler1);
    form.instance().formInspector.addFinishFieldHandler(handler2);
    form.find('SortableBlocks').prop('onExpand')(0);
    form.update();
    form.find('SortableBlock Field').at(0).instance().handleFinish();
    expect(handler1).toHaveBeenLastCalledWith('/block/0/text', '/block/types/default/form/text');
    expect(handler2).toHaveBeenLastCalledWith('/block/0/text', '/block/types/default/form/text');
});

test('Should pass formInspector, schema, data and showAllErrors flag to Renderer', () => {
    const store = new FormStore(new ResourceStore('snippet', '1'));
    // $FlowFixMe
    store.schema = {};
    store.data.title = 'Title';
    store.data.description = 'Description';
    const form = shallow(<Form onSubmit={jest.fn()} store={store} />);

    expect(form.find('Renderer').props()).toEqual(expect.objectContaining({
        data: store.data,
        schema: store.schema,
    }));

    const formInspector = form.find('Renderer').prop('formInspector');
    expect(formInspector.resourceKey).toEqual('snippet');
    expect(formInspector.id).toEqual('1');
});

test('Should pass showAllErrors flag to Renderer when form has been submitted', () => {
    const store = new FormStore(new ResourceStore('snippet', '1'));
    const form = mount(<Form onSubmit={jest.fn()} store={store} />);

    expect(form.find('Renderer').prop('showAllErrors')).toEqual(false);
    form.find(Form).instance().submit();
    form.update();
    expect(form.find('Renderer').prop('showAllErrors')).toEqual(true);
});

test('Should change data on store when changed', () => {
    const submitSpy = jest.fn();
    const store = new FormStore(new ResourceStore('snippet', '1'));
    const form = shallow(<Form onSubmit={submitSpy} store={store} />);

    form.find('Renderer').simulate('change', 'field', 'value');
    expect(store.change).toBeCalledWith('field', 'value');
});

test('Should change data on store without sections', () => {
    const submitSpy = jest.fn();
    const store = new FormStore(new ResourceStore('snippet', '1'));
    // $FlowFixMe
    store.schema = {
        section1: {
            label: 'Section 1',
            type: 'section',
            items: {
                item11: {
                    label: 'Item 1.1',
                    type: 'text_line',
                },
                section11: {
                    label: 'Section 1.1',
                    type: 'section',
                },
            },
        },
        section2: {
            label: 'Section 2',
            type: 'section',
            items: {
                item21: {
                    label: 'Item 2.1',
                    type: 'text_line',
                },
            },
        },
    };

    const form = mount(<Form onSubmit={submitSpy} store={store} />);
    form.find('Input').at(0).instance().handleChange({currentTarget: {value: 'value!'}});

    expect(store.change).toBeCalledWith('item11', 'value!');
});

test('Should show a GhostDialog if the current locale is not translated', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('de')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog').prop('open')).toEqual(true);
});

test('Should not show a GhostDialog if the current locale is translated', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('en')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog').prop('open')).toEqual(false);
});

test('Should show a GhostDialog after the locale has been switched to a non-translated one', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('en')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog').prop('open')).toEqual(false);

    resourceStore.setLocale('de');

    form.update();
    expect(form.find('GhostDialog').prop('open')).toEqual(true);
});

test('Should not show a GhostDialog if the entity does not exist yet', () => {
    const resourceStore = new ResourceStore('snippet', undefined, {locale: observable.box('en')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog')).toHaveLength(0);
});

test('Should not show a GhostDialog if the entity is not translatable', () => {
    const resourceStore = new ResourceStore('snippet', '1');
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog')).toHaveLength(0);
});

test('Should show a GhostDialog and copy the content if the confirm button is clicked', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('de')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog').prop('open')).toEqual(true);
    form.find('GhostDialog Button[skin="primary"]').simulate('click');
    expect(form.find('GhostDialog').prop('open')).toEqual(false);

    expect(formStore.copyFromLocale).toBeCalledWith('en');
});

test('Should show a GhostDialog and do nothing if the cancel button is clicked', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('de')});
    resourceStore.data.availableLocales = ['en'];
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.find('GhostDialog').prop('open')).toEqual(true);
    form.find('GhostDialog Button[skin="secondary"]').simulate('click');
    expect(form.find('GhostDialog').prop('open')).toEqual(false);

    expect(formStore.copyFromLocale).not.toBeCalled();
});

test('Should not show a GhostDialog if the resourceStore is currently loading', () => {
    const resourceStore = new ResourceStore('snippet', '1', {locale: observable.box('de')});
    resourceStore.data.availableLocales = ['en'];
    resourceStore.loading = true;
    const formStore = new FormStore(resourceStore);
    const form = mount(<Form onSubmit={jest.fn()} store={formStore} />);

    expect(form.instance().displayGhostDialog).toEqual(false);
});
