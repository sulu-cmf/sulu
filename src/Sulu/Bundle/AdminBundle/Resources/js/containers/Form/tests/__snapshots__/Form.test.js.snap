// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render form using renderer 1`] = `
<form>
  <div
    class="grid"
  />
</form>
`;
